src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVFormatter.scala

```diff
@@ -11,102 +10,98 @@ import spray.json.JsValue
-              case _ => IndexedSeq.empty
-            }
+          case (attributeName, _) => attributeName.equals(AttributeName.withDefaultNS(collectionMemberType))
+        }.flatMap {
+          case (_, AttributeEntityReference(entityType, entityName)) => Seq(IndexedSeq[String](entity.name, entityName))
+          case (_, AttributeEntityReferenceList(refs)) => refs.map( ref => IndexedSeq[String](entity.name, ref.entityName) )
+          case _ => Seq.empty
         }
     }
     exportToString(headers, rows.toIndexedSeq)
   }
 
-  def makeEntityTsvString(entities: List[EntityWithType], entityType: String): String = {
+  def makeEntityTsvString(entities: Seq [RawlsEntity], entityType: String, requestedHeaders: Option[IndexedSeq[String]]): String = {
     val headerRenamingMap: Map[String, String] = ModelSchema.getAttributeExportRenamingMap(entityType).getOrElse(Map.empty[String, String])
     val entityHeader = "entity:" + entityType + "_id"
     // if we have a set entity, we need to filter out the attribute array of the members so that we only
     // have top-level attributes to construct columns from.
     val memberType = ModelSchema.getCollectionMemberType(entityType)
     val filteredEntities = memberType match {
-      case Success(collectionType) if collectionType.isDefined =>
-        ModelSchema.getPlural(collectionType.get) match {
+      case Success(Some(collectionType)) =>
+        ModelSchema.getPlural(collectionType) match {
           case Success(attributeName) => filterAttributeFromEntities(entities, attributeName)
           case _ => entities
         }
       case _ => entities
     }
 
-    val headers: immutable.IndexedSeq[String] = entityHeader +: filteredEntities.
-      filter { _.entityType == entityType }.
-      map { _.attributes.getOrElse(Map.empty) }.
-      flatMap(_.keySet).
-      distinct.
-      map { key => headerRenamingMap.getOrElse(key, key) }.
-      toIndexedSeq
-    val rows: immutable.IndexedSeq[IndexedSeq[String]] = filteredEntities.filter { _.entityType == entityType }
+
+    val requestedHeadersSansId = requestedHeaders.
+      // remove empty strings
+      map(_.filter(_.length > 0)).
+      // handle empty requested headers as no requested headers
+      flatMap(rh => if (rh.isEmpty) None else Option(rh)).
+      // entity id always needs to be first and is handled differently so remove it from requestedHeaders
+      map(_.filterNot(_.equalsIgnoreCase(entityType + "_id")))
+
+    val headers: IndexedSeq[String] = entityHeader +: requestedHeadersSansId.getOrElse(defaultHeaders(entityType, filteredEntities, headerRenamingMap))
+    val rows: IndexedSeq[IndexedSeq[String]] = filteredEntities.filter { _.entityType == entityType }
       .map { entity =>
         makeRow(entity, headers, headerRenamingMap)
       }.toIndexedSeq
     exportToString(headers, rows)
   }
 
+  def defaultHeaders(entityType: String, filteredEntities: Seq[RawlsEntity], headerRenamingMap: Map[String, String]) = {
```

Comment:
Nice extracted function and simplification. I'll have to remember to keep `collect` in mind.